### DISCUSSION TITLE
GitHub Actions

### DISCUSSION THREAD

**Discussion by michaliskambi** [pasdoc/pasdoc#157] [Announcements]
I started work on GitHub Actions for PasDoc.

Every time you push, to whatever PR / branch, GH Actions will check your commit now.

Currently running checks are on https://github.com/pasdoc/pasdoc/actions .

The configuration what happens is in https://github.com/pasdoc/pasdoc/blob/master/.github/workflows/build.yml file. Right now it builds and packages PasDoc (command-line, tools, GUI) on Windows and Linux. More to come, in particular: `make tests` should eventually be done by GH Actions.

**Comment by michaliskambi** [comment]
Note: I noticed that sometimes Lazarus unfortunately just fails with `Access Violation`, and it seems to be random. The solution is just to re-run the build (you can ask to "Re-run failed jobs" if you navigate to the proper action on https://github.com/pasdoc/pasdoc/actions ).

Maybe I'll just change the script to retry Lazarus once, always, in shell script this is easy:

```
if ! lazbuild "${LPI_FILENAME}"; then
  echo '1st execution of lazbuild failed, trying again'
  lazbuild "${LPI_FILENAME}"
fi
```

**Reply by michaliskambi** [reply]
Basic GitHub Actions setup on PasDoc `master` branch (so, also all branches you make from it since now) is now working:

- We build (`make`, `make build-tools`, `make dist-xxx`) on Linux, Windows, macOS
- We do `make tests` too (for now only on Linux - for some reason "diff -ur ..." doesn't work OK on Windows). 
    * XML and PHP validation of output is done too, the runner has all necessary programs. 
    * HTML validation can be done too but is commented out for now -- it reports (correctly) failures which we should fix, see https://github.com/pasdoc/pasdoc/blob/master/.github/workflows/build.yml#L54 for details.
- P.S. lazbuild is now executed twice (to retry, in case it failed with random _Access Violation_)

All the branches, PRs, including in all forks, will now enjoy all automated checks :)

![screenshot 2022-11-17 19:01:16 select](https://user-images.githubusercontent.com/1076978/202523841-2eebaa8f-4a58-4bab-a34e-8bc4581fe08c.png)

**Reply by Fr0sT-Brutal** [reply]
> * for some reason "diff -ur ..." doesn't work OK on Windows)

Maybe use git diff just like in run_all_tests.sh?

**Reply by michaliskambi** [reply]
That diff call is there (in `check_cache.sh`) necessary to compare directories that are not part of the repo. They are only temporarily created, compared and later removed by `check_cache.sh`.

So we do need a regular `diff`. 

We use `git diff` in other place when doing tests, to make sure that new test output matches what is in repo (`git diff -w --exit-code testcases_output/`). But that's a different usecase :)
